File: scripts/compute_daily_metrics.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

function loadJson(p) {
  try {
    const full = path.resolve(p);
    if (!fs.existsSync(full)) return [];
    return JSON.parse(fs.readFileSync(full, 'utf8'));
  } catch (e) {
    console.error('Error loading', p, e.message);
    return [];
  }
}

function extractOrderCounts(rec) {
  for (const k of ['order_counts', 'customer_count', 'customer_count_data']) {
    if (rec && Object.prototype.hasOwnProperty.call(rec, k)) {
      const v = rec[k];
      if (v && typeof v === 'object') return v;
    }
  }
  return {};
}

function numberOrZero(v) {
  if (v === null || v === undefined) return 0;
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
}

function buildIndex(arr, keyExtractor) {
  const out = Object.create(null);
  for (const r of arr || []) {
    const d = r && r.date;
    if (!d) continue;
    out[d] = keyExtractor ? keyExtractor(r) : r;
  }
  return out;
}

function main(options = {}) {
  const sales = loadJson(options.sales || 'Dashboard_data.store_wise_sales.json');
  const payroll = loadJson(options.payroll || 'Payroll_dashboard.Payroll_daily.json');
  const waste = loadJson(options.waste || 'Dashboard_data.waste_tool.json');
  const gross = loadJson(options.gross || 'Dashboard_data.gross_margin.json');
  const customer = loadJson(options.customer || 'Dashboard_data.customer_count.json');

  const salesByDate = buildIndex(sales, r => r.sales_data || {});
  const payrollByDate = buildIndex(payroll, r => r.payroll_data || {});
  const wasteByDate = buildIndex(waste, r => r.total_waste_per_store || {});
  const grossByDate = buildIndex(gross, r => r.stores || {});
  const custByDate = buildIndex(customer, r => extractOrderCounts(r));

  const allDates = Array.from(new Set([
    ...Object.keys(salesByDate),
    ...Object.keys(payrollByDate),
    ...Object.keys(wasteByDate),
    ...Object.keys(grossByDate),
    ...Object.keys(custByDate),
  ])).sort();

  const results = [];

  for (const date of allDates) {
    const sdata = salesByDate[date] || {};
    const pdata = payrollByDate[date] || {};
    const wdata = wasteByDate[date] || {};
    const gdata = grossByDate[date] || {};
    const cdata = custByDate[date] || {};

    const stores = new Set([
      ...Object.keys(sdata),
      ...Object.keys(pdata),
      ...Object.keys(wdata),
      ...Object.keys(gdata),
      ...Object.keys(cdata),
    ]);

    const dataSection = {};
    for (const store of Array.from(stores).sort()) {
      const srec = sdata[store] || {};
      const net_sales = numberOrZero(srec.net_sales);
      const total_waste = numberOrZero(wdata[store]);
      const net_revenue = net_sales - total_waste;

      let payrollVal = pdata[store];
      if (payrollVal === null || payrollVal === undefined) payrollVal = 0;
      payrollVal = numberOrZero(payrollVal);

      const net_revenue_payroll_adjusted = net_revenue - payrollVal;

      const g = gdata[store] || {};
      const total_revenue = numberOrZero(g.total_revenue);
      const total_cost = numberOrZero(g.total_cost);
      let gross_margin_percent = (g && g.gross_margin_percent !== undefined) ? g.gross_margin_percent : null;
      if ((gross_margin_percent === null || gross_margin_percent === undefined) && total_revenue) {
        gross_margin_percent = ((total_revenue - total_cost) / total_revenue) * 100;
      }

      let gm_payroll_adj = null;
      if (total_revenue) {
        gm_payroll_adj = ((total_revenue - total_cost - payrollVal) / total_revenue) * 100;
      }

      let orders = cdata[store];
      orders = numberOrZero(orders);

      let avg_basket = null;
      if (orders) avg_basket = net_revenue / orders;

      dataSection[store] = {
        Net_Revenue: Number(net_revenue.toFixed(2)),
        Net_Revenue_payroll_Adjusted: Number(net_revenue_payroll_adjusted.toFixed(2)),
        'GM%': (gross_margin_percent !== null && gross_margin_percent !== undefined) ? Number(gross_margin_percent.toFixed(2)) : null,
        'GM%_Payroll_Adjusted': (gm_payroll_adj !== null && gm_payroll_adj !== undefined) ? Number(gm_payroll_adj.toFixed(2)) : null,
        Average_Basket: (avg_basket !== null && avg_basket !== undefined) ? Number(avg_basket.toFixed(2)) : null,
      };
    }

    results.push({
      _id: { $oid: crypto.randomBytes(12).toString('hex') },
      date,
      data: dataSection,
      updated_at: { $date: new Date().toISOString() },
    });
  }

  const outPath = options.out || 'daily_metricsjs.json';
  fs.writeFileSync(outPath, JSON.stringify(results, null, 2), 'utf8');
  console.log(`Wrote ${results.length} records to ${outPath}`);
}

if (require.main === module) {
  const out = process.argv[2];
  main({ out });
}

File: scripts/compute_daily_metrics.py
import json
import sys
from pathlib import Path
from datetime import datetime
import uuid


def load_json(path):
    if not Path(path).exists():
        return []
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def extract_order_counts(rec):
    # try common keys used in provided files
    for k in ('order_counts', 'customer_count', 'customer_count_data', 'customer_count_data'):
        v = rec.get(k)
        if isinstance(v, dict):
            return v
    # sometimes order_counts is nested or missing
    return {}


def main(
    sales_path='Dashboard_data.store_wise_sales.json',
    payroll_path='Payroll_dashboard.Payroll_daily.json',
    waste_path='Dashboard_data.waste_tool.json',
    gross_path='Dashboard_data.gross_margin.json',
    customer_path='Dashboard_data.customer_count.json',
    out_path='daily_metrics.json',
):
    sales = load_json(sales_path)
    payroll = load_json(payroll_path)
    waste = load_json(waste_path)
    gross = load_json(gross_path)
    customer = load_json(customer_path)

    # index by date
    sales_by_date = {r.get('date'): r.get('sales_data', {}) for r in sales}
    payroll_by_date = {r.get('date'): r.get('payroll_data', {}) for r in payroll}
    waste_by_date = {r.get('date'): r.get('total_waste_per_store', {}) for r in waste}
    gross_by_date = {r.get('date'): r.get('stores', {}) for r in gross}
    cust_by_date = {r.get('date'): extract_order_counts(r) for r in customer}

    all_dates = sorted(set(list(sales_by_date) + list(payroll_by_date) + list(waste_by_date) + list(gross_by_date) + list(cust_by_date)))

    results = []
    for date in all_dates:
        sales_data = sales_by_date.get(date, {}) or {}
        payroll_data = payroll_by_date.get(date, {}) or {}
        waste_data = waste_by_date.get(date, {}) or {}
        gross_data = gross_by_date.get(date, {}) or {}
        cust_data = cust_by_date.get(date, {}) or {}

        stores = set()
        stores.update(sales_data.keys())
        stores.update(payroll_data.keys())
        stores.update(waste_data.keys())
        stores.update(gross_data.keys())
        stores.update(cust_data.keys())

        data_section = {}
        for store in sorted(stores):
            # net_sales from sales_data
            net_sales = 0.0
            srec = sales_data.get(store)
            if isinstance(srec, dict):
                net_sales = srec.get('net_sales', 0.0) or 0.0

            total_waste = waste_data.get(store, 0.0) or 0.0

            net_revenue = net_sales - total_waste

            payroll_val = payroll_data.get(store, 0.0) or 0.0
            # payroll entries may be None
            if payroll_val is None:
                payroll_val = 0.0

            net_revenue_payroll_adjusted = net_revenue - payroll_val

            g = gross_data.get(store, {}) or {}
            total_revenue = g.get('total_revenue') or 0.0
            total_cost = g.get('total_cost') or 0.0
            gross_margin_percent = g.get('gross_margin_percent')
            if gross_margin_percent is None:
                # compute if possible
                if total_revenue:
                    gross_margin_percent = (total_revenue - total_cost) / total_revenue * 100
                else:
                    gross_margin_percent = None

            # GM% payroll adjusted
            gm_payroll_adj = None
            if total_revenue:
                gm_payroll_adj = (total_revenue - total_cost - payroll_val) / total_revenue * 100

            # order counts
            orders = cust_data.get(store)
            try:
                orders = int(orders) if orders is not None else 0
            except Exception:
                orders = 0

            avg_basket = None
            if orders:
                avg_basket = net_revenue / orders

            # round/keep floats consistent
            entry = {
                'Net_Revenue': round(net_revenue, 2),
                'Net_Revenue_payroll_Adjusted': round(net_revenue_payroll_adjusted, 2),
                'GM%': round(gross_margin_percent, 2) if gross_margin_percent is not None else None,
                'GM%_Payroll_Adjusted': round(gm_payroll_adj, 2) if gm_payroll_adj is not None else None,
                'Average_Basket': round(avg_basket, 2) if avg_basket is not None else None,
            }
            data_section[store] = entry

        out_rec = {
            '_id': {'$oid': uuid.uuid4().hex[:24]},
            'date': date,
            'data': data_section,
            'updated_at': {'$date': datetime.utcnow().isoformat() + 'Z'},
        }
        results.append(out_rec)

    # write output
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(results, f, indent=2)

    print(f'Wrote {len(results)} date records to {out_path}')


if __name__ == '__main__':
    args = sys.argv[1:]
    kwargs = {}
    if args:
        # allow specifying output path as first arg
        kwargs['out_path'] = args[0]
    main(**kwargs)
